### ISSUE TITLE
Tacoma ranch should be a unique location

### ISSUE THREAD

**Issue by Vulp0d** [CleverRaven/Cataclysm-DDA#76921]
### Is your feature request related to a problem? Please describe.

I have found 3 Tacoma ranches in a close proximity, which is weird, considering that other faction locations are unique.

### Solution you would like.

Since other faction locations are unique, it would make sense if this one would be unique as well.

### Describe alternatives you have considered.

N/A

### Additional context

N/A

**Comment by PatrikLundell** [comment]
Tacoma is different from the others in that it's big and unprotected. This means that it's very common to have area threats right adjacent to it, such as fungus and ants, and have those threats start to pour in as soon as it enters the reality bubble.
The refugee center is also large, but most of it acts as a buffer against such threats.

However, it currently doesn't help to have multiple Tacomas, since the Refugee Center selects one without any evaluation of how besieged it is.

**Comment by IdleSol** [comment]
I've always wondered, is it possible to make a tag for the map generator that no monsters can be created in this tile? And can it be applied by condition within a radius of the center.

This is to the question of monsters near important locations.

**Comment by PatrikLundell** [comment]
As far as I understand, there is some rudimentary monster site avoidance immediately adjacent to such sites available, but that's not really cutting it for dealing with area threats such as ants and fungii. The system blocks the monster areas from infesting the sites themselves, so they start out free of monsters, but that doesn't do anything to keep them from starting to pour into the sites once the reality bubble comes into play.

Hm, a thought: Monster area threats mark tiles around them as "infected". Would it be possible for important sites to do the opposite, i.e. clear out a radius around them from area threats, and not just remove/block these tags from the sites themselves?

**Comment by IdleSol** [comment]
Cycle the world generator with different parameters....

1. Land generation
2. Overlay oceans, mark boundaries with label, coastline. Add indications of which side is adjacent and/or additional weights. Depth 2 map cells (omt?)
3. Overlay rivers and lakes, label boundaries with tags river/lake shore.
4. Overlay forest, label with forest tags. Cells labeled water, block the forest overlay. 
5. Plant seeds of cities, add size parameters to them (maybe type?).
6. Grow a city from seeds. Perform interactions with labels from past iterations (label water = bridge, label coast = bungalow, etc.)
7. Connect cities with roads. There was a PR on highways, but don't remember what became of it. It would be good for cities larger than X size.
8. By analogy, add everything else except the unique ones. Attach them to the nearest road.
9. Add unique locations. Check the conditions for objects that are connected to each other. The same ranch should be near the refugee center. So, having chosen a place under the center, we need to check the conditions for the possibility of placing the ranch. If the condition is not met, look for a new location. We add areas of ban on the cataclysm. Depth X cells from the center. If necessary, an array of depth X + prohibition of generation Y. (In a radius of two cells, can not be ants, and in a radius of 10 cells can not be mushrooms)
10. We get a ready map before the cataclysm. (We use it to display when reading guides and everything that can open the map, but does not offer knowledge about the consequences of the cataclysm). 
11. Each tile has a set of labels and weights. We make a cataclysm and spread mushrooms, ants, and migos. Bomb cities and stuff.

Basically looking at the map generation in dwarf fortress.

P.S. Once, the starting location was a place near the triffids. From memory 2 or 3 squares from their location. Normally there would be a lot of mobs at that distance. But there weren't any in range. Perhaps there is some kind of defense?

**Comment by Zireael07** [comment]
The problem is not making a map, the problem is ensuring those unique locations do not get wiped by a monster wave

**Comment by PatrikLundell** [comment]
The Dwarf Fortress approach differs very significantly from the C:DDA one, and a conversion would be quite tricky. I know the general DF process fairly well, but only have spotty knowledge of the C:DDA one.

DF generates everything at world generation, where "everything" is basic parameters plus a set of seeds for random generators, so the actual data stored is fairly limited. The actual mid level tiles are only generated from this when needed, and and discarded when no longer needed, with only changes made being stored and then applied when the tiles are regenerated.

C:DDA uses a generation on first use approach, which means that when an Overmap Tile (OMT, 24*24 in-game "1 m" tiles) is touched by the reality bubble it's recovered from memory if loaded previously in the session, loaded from file if it has been generated previously, but not in the current session, and generated if no file is found either. The generated data is produced using random factors without any seed or random number generator info kept, so you get different variants and outcomes from different attempts (save the game, move into unknown terrain, look at the terrain, quit without saving, load, and move into the same terrain again, and it will be different). I don't know when the overmaps (huge blocks of OMTs, most easily visualized by using debug reveal, as that shows you the boundaries on the overmap) are generated (when touched or at world gen), although I suspect only some basic data is generated at world gen.

Also note that you COULD place unique locations at world gen, but that means you'd get difficulties getting start locations and the ones supposed to be nearby to line up (and I have no idea how this even works when you start with new characters in an existing world).

Yes, there is some kind of defense. I've looked at a Tacoma farm with mycus on three sides, and debug overmap data showed infestation in all the surrounding tiles, but all the Tacoma ones free of it. However, you'd need a defense zone around the facilities, and those zones would have to be fairly large, because mycus and ant radii are large, so a two tile buffer would still surround Tacoma if an ant or mycus center location was right next to this buffer.

**Comment by IdleSol** [comment]
I may have used the terms incorrectly. I am talking about `overmap`, which is 180x180 `omt`. Not about `omt` = 24x24 tiles.

When the game generates an `overmap`, then place records of what is in the `omt`. For example, a house or a lake or a mushroom tower. 

If you do this not in one pass of the generator, but in several. You can get interesting results. Especially if the next cycle of the generator will take into account the previous cycle and special labels.

According to OVERMAP.md. Map generation function, it is a sequential function call.
```
populate_connections_out_from_neighbors( north, east, south, west );
place_rivers( north, east, south, west );
place_lakes();
place_forests();
place_swamps();
place_ravines();
place_cities();
place_forest_trails();
place_roads( north, east, south, west );
place_specials( enabled_specials );
place_forest_trailheads();
polish_river();
place_mongroups();
place_radios();
```

Perhaps all we need to do is change the order in which these functions are called. And teach them to work with new labels. It may be necessary to separate `omt` related to the cataclysm into a separate function. Which will be the last in the call queue.

Quick tests showed that a minimum of two neighboring overmaps, generated at game start. Create a character - save the game - teleport to a neighboring map x2, open the map - remember it. Exit without saving and repeat teleports in the same direction. I got the same town twice (looking at the name) in roughly the same place. But I didn't do the cell-by-cell comparisons.

**Comment by PatrikLundell** [comment]
I would expect all the game important locations to be specials, but that's a category that includes most of the interesting stuff on the map. Ant hills are included in this category, but wasp nests are not, while there is one dermatik nest that's a special and one that's an mx_special (which are generated on OMT generation).

Note that for generation to separate these kinds of specials you'd have to mark them separately. You could of course use the existing unique flag(s) for that purpose (and you may or may not want overmap uniques to be included).
You can do it in any order, i.e. place important locations first and then have logic that places monster specials such that they're never too close, or the monster sites first and important sites later, avoiding monster zones. Regardless of the order, you'd have to identify monster sites and extract their monster generation radii to perform this separation operation. How? I don't know. It may require some new tag, and it might contribute to moving generation code to JSON to expose the radii if access to that is missing (I guess you could throw e.g. a nuclear contamination area into such a category as well, should such a thing be introduced in the future).

I would probably go for placement of important stuff first, because it means it's almost guaranteed to be placed, leading to less pushing off of it to hopefully be generated in another OM, but doing that to overmap_uniques would probably require an adjustment of their generation rates, as you'd end up with more of them due to lack of placement competition. Also, this may affect "far away" important location probabilities, resulting in a need to adjust those as well, or the near guaranteed placement may result in them ending up too close.

And, of course, there may be other considerations in the placement code that would be overturned by a separation, and thus dealt with.

**Comment by esotericist** [comment]
@PatrikLundell should this be closed as of #77276 ?

**Comment by PatrikLundell** [comment]
Yes. The PR really should have closed it, as the request itself specifies Tacoma and nothing else. The following discussion introduced other elements, but that's not really the request itself.

The task of safe guarding access to facilities remain, and the measures introduced in the PR to protect facilities aren't complete either. But still, Tacoma was made globally unique, which is what the request itself is about.
